app/src/test/java/org/simple/clinic/security/pin/PinEntryCardControllerTest.kt

```diff
@@ -6,36 +6,64 @@ import com.nhaarman.mockito_kotlin.mock
 import com.nhaarman.mockito_kotlin.times
 import com.nhaarman.mockito_kotlin.verify
 import com.nhaarman.mockito_kotlin.whenever
+import io.reactivex.Completable
 import io.reactivex.Observable
 import io.reactivex.Single
+import io.reactivex.plugins.RxJavaPlugins
+import io.reactivex.schedulers.TestScheduler
 import io.reactivex.subjects.PublishSubject
+import junitparams.JUnitParamsRunner
+import junitparams.Parameters
+import org.junit.After
 import org.junit.Before
 import org.junit.Test
+import org.junit.runner.RunWith
 import org.simple.clinic.patient.PatientMocker
 import org.simple.clinic.security.ComparisonResult
 import org.simple.clinic.security.PasswordHasher
+import org.simple.clinic.security.pin.BruteForceProtection.ProtectedState
+import org.simple.clinic.security.pin.PinEntryCardView.State
 import org.simple.clinic.user.UserSession
 import org.simple.clinic.widgets.UiEvent
+import org.threeten.bp.Clock
+import org.threeten.bp.Duration
+import org.threeten.bp.Instant
+import org.threeten.bp.ZoneOffset.UTC
 
+@RunWith(JUnitParamsRunner::class)
 class PinEntryCardControllerTest {
 
   private val screen = mock<PinEntryCardView>()
   private val userSession = mock<UserSession>()
   private val passwordHasher = mock<PasswordHasher>()
+  private val bruteForceProtection = mock<BruteForceProtection>()
 
   private lateinit var controller: PinEntryCardController
   private val uiEvents = PublishSubject.create<UiEvent>()
-
-  val loggedInUser = PatientMocker.loggedInUser()
+  private val clock = Clock.fixed(Instant.now(), UTC)
+  private val loggedInUser = PatientMocker.loggedInUser()
+  private val testScheduler = TestScheduler()
 
   @Before
   fun setUp() {
     whenever(userSession.requireLoggedInUser()).thenReturn(Observable.just(loggedInUser))
     whenever(passwordHasher.compare(any(), any())).thenReturn(Single.never<ComparisonResult>())
+    whenever(bruteForceProtection.incrementFailedAttempt()).thenReturn(Completable.complete())
+    whenever(bruteForceProtection.recordSuccessfulAuthentication()).thenReturn(Completable.complete())
+    whenever(bruteForceProtection.protectedStateChanges()).thenReturn(Observable.never())
+
+    RxJavaPlugins.setComputationSchedulerHandler { testScheduler }
```

Comment:
I feel that in these situations, we might need to add a comment of some sort as to why we are overriding this specific scheduler because we don't explicitly use this scheduler in our thread.

Otherwise, it's not very evident that this is needed because `Observable.timer` internally runs on the computation scheduler.

Maybe we should consider injecting the scheduler `Observable.timer` runs on manually as well.

Reply:
> Otherwise, it's not very evident that this is needed because Observable.timer internally runs on the computation scheduler.

Won't that be documenting the expected behavior of delay based operators in RxJava? They all run on computation schedulers.

Reply:
Hold on, `TestScheduler` isn't being used anymore in this test. Problem solved!